src/main/java/onlinebookstore/service/order/OrderServiceImpl.java

```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(OrderShippingAddressRequestDto requestDto, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUser(user).orElseThrow(() ->
+                new EntityNotFoundException("Can't find shopping cart by user: " + user));
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(orderItemMapper::toModelFromCartItem)
+                .collect(Collectors.toSet());
+        BigDecimal total = shoppingCart.getCartItems().stream()
+                .map(ci -> ci.getBook().getPrice().multiply(BigDecimal.valueOf(ci.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setShippingAddress(requestDto.shippingAddress());
+        order.setOrderDate(LocalDateTime.now());
+        order.setTotal(total);
+        order = orderRepository.save(order);
+        addOrderItems(order, orderItems);
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderDto> getAllOrders(User user, Pageable pageable) {
+        return orderRepository.findAllByUser(user, pageable).stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public List<OrderItemDto> getAllOrderItemsByOrderId(Long orderId,
+                                                        User user,
+                                                        Pageable pageable) {
+        return orderRepository.getAllOrderItemsByOrderId(orderId, user, pageable).stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public OrderItemDto getOrderItemByOrderIdAndItemId(Long orderId,
+                                                       Long itemId,
+                                                       User user) {
+        Order order = orderRepository.findOrderByIdAndUser(orderId, user).orElseThrow(
+                () -> new EntityNotFoundException("Can't find order by id : " + orderId
+                + " and by user: " + user)
+        );
+        OrderItem orderItem = order.getOrderItems().stream()
+                .filter(oi -> Objects.equals(oi.getId(), itemId))
+                .findFirst()
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can't find item by id: " + itemId));
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    @Override
+    public void updateStatus(Long orderId, OrderStatusRequestDto requestDto) {
+        orderRepository.updateStatus(orderId, requestDto.status());
+    }
+
+    private void addOrderItems(Order order, Set<OrderItem> orderItems) {
+        Set<OrderItem> orderItemSet = orderItems.stream()
+                .peek(oi -> oi.setOrder(order)).collect(Collectors.toSet());

```

Comment:
Also learn what is peek. It doesnt change the Stream. So Set<OrderItem> orderItemSet won't have order set. Use forEach instead.